### PR TITLE
Follow-up for Camera TLS Certificate Management Cluster to bring it up to latest spec

### PR DESCRIPTION
**Pull request by gmarcosb**
#### Summary

This is a follow-up to https://github.com/project-chip/connectedhomeip/pull/39754:

* Updates to latest spec up-to & including https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/12201
* Flesh out existing TC test scripts (2.2 & 2.9) to match https://github.com/CHIP-Specifications/chip-test-plans/pull/5447

#### Testing

CI; requires https://github.com/project-chip/connectedhomeip/pull/40817 to pass

### CODE REVIEW

**Review by samadDotDev — COMMENTED** [review]
Can we also bring the TLS client management cluster upto the same spec version so we match the corresponding cert management updates here? Or are you planning to do that as a follow-on to this follow-up PR? :)

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter
Comment (samadDotDev) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1]: Might also need to re-run codegen after the merge of https://github.com/project-chip/connectedhomeip/pull/40549 so that these changes are reflected in the camera app as well - I'm not sure if CI or other conflicts would enforce that update otherwise.

**Comment by gmarcosb** [comment]
> Can we also bring the TLS client management cluster upto the same spec version so we match the corresponding cert management updates here? Or are you planning to do that as a follow-on to this follow-up PR? :)

separate PR (not follow-up, parallel) - the PRs get big enough on their own, & splitting it on the cluster boundary is a good way to parallelize without making big PRs 😅🤣

File: src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml [hunk 1]
```diff
@@ -51,107 +46,109 @@ Git: 0.9.2-summer2025-41-g99efd0498
...
-    <attribute code="0x0003" side="server" define="PROVISIONED_CLIENT_CERTIFICATES" type="array" entryType="TLSClientCertificateDetailStruct" length="254">ProvisionedClientCertificates</attribute>
-    <command code="0x00" source="client" name="ProvisionRootCertificate" optional="false" response="ProvisionRootCertificateResponse" isFabricScoped="true">
+    <attribute code="0x0000" side="server" name="MaxRootCertificates" define="MAX_ROOT_CERTIFICATES" type="int8u" min="5" max="254"/>
+    <attribute code="0x0001" side="server" name="ProvisionedRootCertificates" define="PROVISIONED_ROOT_CERTIFICATES" type="array" entryType="TLSCertStruct" length="254"/>
+    <attribute code="0x0002" side="server" name="MaxClientCertificates" define="MAX_CLIENT_CERTIFICATES" type="int8u" min="2" max="254"/>
+    <attribute code="0x0003" side="server" name="ProvisionedClientCertificates" define="PROVISIONED_CLIENT_CERTIFICATES" type="array" entryType="TLSClientCertificateDetailStruct" length="254"/>
+    <command code="0x00" source="client" name="ProvisionRootCertificate" response="ProvisionRootCertificateResponse" isFabricScoped="true">
```
Comment (cecille) [inline src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:53]: does zap differentiate between fabric scoped and fabric sensitive? Because in the spec, these are marked as fabric sensitive. https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/tls/TLSCertificateManagement.adoc#5-attributes

File: examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp [hunk 2]
```diff
@@ -92,8 +92,14 @@ struct RefEncodableClientCert
     {
         VerifyOrReturnError(certificate != nullptr, CHIP_ERROR_INTERNAL);
 
-        MutableByteSpan targetBytes(certBytes);
-        ReturnErrorOnFailure(CopySpanToMutableSpan(src.clientCertificate.Value(), targetBytes));
+        if (!src.clientCertificate.Value().IsNull())
```
Comment (cecille) [inline examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp:95]: General Q about this file - why is so much of the cluster logic implemented in the example? This seems like logic that should be part of the cluster implementation so other implementers can inherit it. Does the camera example have a similar implementation? This PR doesn't have the same changes there - why is that example different?

**Review by marktrayer — APPROVED** [review]

File: src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include <stdlib.h>
 
 using namespace chip;
+using namespace chip::app::DataModel;
```
Comment (cecille) [inline src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp:25]: general follow up - this seems like this would be unit testable. Please add tests. I think follow up might be reasonable here given that there are cert tests in the PR, but there are certainly corner cases in this code that are not hit by the cert tests.

File: src/python_testing/TC_TLSCERT.py [hunk 4]
```diff
@@ -53,24 +54,45 @@
...
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        # See https://github.com/project-chip/connectedhomeip/issues/39690
-        self.skip_cr2 = True
 
-    def gen_cert(self) -> bytes:
+    def get_key(self) -> rsa.RSAPrivateKey:
         # Start generating the root certificate
-        root_key = rsa.generate_private_key(
+        return rsa.generate_private_key(
```
Comment (cecille) [inline src/python_testing/TC_TLSCERT.py:74]: why is this returning a new key if you have self.key already?

File: src/python_testing/TC_TLSCERT.py [hunk 5]
```diff
@@ -53,24 +54,45 @@
...
+            self.key = key
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        # See https://github.com/project-chip/connectedhomeip/issues/39690
-        self.skip_cr2 = True
 
-    def gen_cert(self) -> bytes:
+    def get_key(self) -> rsa.RSAPrivateKey:
         # Start generating the root certificate
```
Comment (cecille) [inline src/python_testing/TC_TLSCERT.py]: comment is now wrong, but github wont' let me edit the line because it's not part of the change set :(

File: src/python_testing/TC_TLSCERT.py [hunk 6]
```diff
@@ -275,295 +290,446 @@ async def read_tls_cert_attribute(self, attribute):
...
         _, noc_resp, _ = await CommissioningBuildingBlocks.AddNOCForNewFabricFromExisting(
             commissionerDevCtrl=cr1, newFabricDevCtrl=cr2,
             existingNodeId=self.dut_node_id, newNodeId=cr2_dut_node_id
         )
         fabric_index_cr2 = noc_resp.fabricIndex
 
+        cr2_cmd = self.TargetedFabric(self, endpoint=endpoint, dev_ctrl=cr2, node_id=cr2_dut_node_id, fabric_index=fabric_index_cr2)
+        return self.TwoFabricData(cr1_cmd=cr1_cmd, cr2_cmd=cr2_cmd)
+
+    def desc_TC_TLSCERT_2_2(self) -> str:
```
Comment (cecille) [inline src/python_testing/TC_TLSCERT.py:401]: is 2.1 implemented anywhere? Is that YAML?

File: src/python_testing/TC_TLSCERT.py [hunk 7]
```diff
@@ -275,295 +290,446 @@ async def read_tls_cert_attribute(self, attribute):
...
+        asserts.assert_equal(find_response.certificateDetails[0].certificate, my_root_cert[2])
+        asserts.assert_equal(find_response.certificateDetails[0].caid, my_caid[1])
+
+        self.step(21)
+        await cr1_cmd.send_remove_root_command(caid=my_caid[0])
+
+        self.step(22)
+        await cr1_cmd.send_remove_fabric_command(cr2_cmd.fabric_index)
+
+    def desc_TC_TLSCERT_2_9(self) -> str:
```
Comment (cecille) [inline src/python_testing/TC_TLSCERT.py:558]: suggest adding pics_ functions for each of the tests too. I know they're marked as only applicable for devices with that cluster on that endpoint, which is actually really helpful for when we start batch-running tests, but the TH still goes by PICS for test selection, so it'll select these tests for every device, then force them to run the test and see that it gets skipped. I know...not ideal. Getting rid of PICS has been a slog and a rant for another day. Anyway, doesn't need to be done in the PR or before the test event, but should add these to give some guidance to the TH.

File: src/python_testing/TC_TLSCERT.py [hunk 8]
```diff
@@ -86,7 +108,7 @@ def gen_cert(self) -> bytes:
             CertificateBuilder()
             .subject_name(root_cert_subject)
```
Comment (cecille) [inline src/python_testing/TC_TLSCERT.py]: this should come from the csr.

File: src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml (hunk 1 above)
Comment (gmarcosb) [reply src/app/zap-templates/zcl/data-model/chip/tls-certificate-management-cluster.xml:53]: Good call-out! Apparently not? 🤔😬 https://github.com/project-chip/zap/blob/d6ee8315f10bbbb20b3edf94763ae647c65e7270/src-electron/db/db-mapping.js#L90

Filed https://github.com/project-chip/zap/issues/1644

File: src/platform/Linux/CHIPLinuxStorage.cpp [hunk 9]
```diff
@@ -230,14 +230,14 @@ CHIP_ERROR ChipLinuxStorage::WriteValueStr(const char * key, const char * val)
 
 CHIP_ERROR ChipLinuxStorage::WriteValueBin(const char * key, const uint8_t * data, size_t dataLen)
 {
-    static const size_t kMaxBlobSize = 5 * 1024;
+    static const size_t kMaxBlobSize = 10 * 1024;
```
Comment (gmarcosb) [inline src/platform/Linux/CHIPLinuxStorage.cpp:233]: Wanted to call out this change - we needed this to be able to store the 10 intermediate certs in the test case, seems reasonable to allow blobs up to 10kb

File: src/app/BufferedReadCallback.cpp [hunk 10]
```diff
@@ -120,7 +120,12 @@ CHIP_ERROR BufferedReadCallback::BufferListItem(TLV::TLVReader & reader)
     handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
     VerifyOrReturnError(!handle.IsNull(), CHIP_ERROR_NO_MEMORY);
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
```
Comment (gmarcosb) [inline src/app/BufferedReadCallback.cpp:123]: Wanted to call out this change as well; reading the attribute over TCP caused a failure here because the buffer wasn't large enough, as such changed this to use chained buffer when we're compiling with TCP support

File: src/python_testing/TC_TLSCERT.py (hunk 6 above)
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py:401]: Not implemented; follow-up PR & tracked @ https://github.com/project-chip/connectedhomeip/issues/40907

File: src/python_testing/TC_TLSCERT.py (hunk 5 above)
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py]: I hate that you can't comment on any line 🤣 thanks, fixed

File: examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp [hunk 11]
```diff
@@ -108,15 +114,12 @@ struct RefEncodableClientCert
             }
             ReturnErrorOnFailure(srcIter.GetStatus());
             certificate->intermediateCertificates.SetValue(List<ByteSpan>(intermediateCerts.data(), i));
-        }
-        else
-        {
-            certificate->intermediateCertificates.ClearValue();
+        } else {
+            certificate->intermediateCertificates.SetValue(List<ByteSpan>());
         }
```
[the remark this answers is not in the document]
Comment (gmarcosb) [reply examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp]: Intentional, storage not having a value is meant to signify an empty list

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp [hunk 12]
```diff
@@ -343,7 +343,8 @@ void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ct
     auto status = mDelegate.GenerateClientCsr(ctx.mRequestPath.mEndpointId, fabric, req, [&](auto & response) -> Status {
         VerifyOrDieWithMsg(response.ccdid <= kMaxClientCertId, NotSpecified, "Spec requires CCDID to be <= kMaxClientCertId");
         VerifyOrDieWithMsg(response.csr.size() <= 3000, NotSpecified, "Spec requires csr.size() to be <= 3000");
-        VerifyOrDieWithMsg(response.nonce.size() <= 128, NotSpecified, "Spec requires nonce.size() to be < kMaxClientCertId");
+        VerifyOrDieWithMsg(response.nonceSignature.size() <= kMaxSignatureBytes, NotSpecified,
+                           "Spec requires nonceSignature.size() to be < kMaxClientCertId");
```
[the remark this answers is not in the document]
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp]: Fixed

File: src/python_testing/TC_TLSCERT.py [hunk 13]
```diff
@@ -275,295 +290,446 @@ async def read_tls_cert_attribute(self, attribute):
...
+        found_certs = await cr2_cmd.send_find_client_command(ccdid=my_ccdid[2])
+        asserts.assert_equal(len(found_certs), 1, "Expected single certificate")
+        asserts.assert_in(my_ccdid[2], found_certs, "FindClientCertificate should contain provisioned client cert")
+        asserts.assert_equal(found_certs[my_ccdid[2]], my_client_cert[2], "Expected matching certificate detail")
+
+        self.step(22)
+        await cr1_cmd.send_provision_client_command(ccdid=my_ccdid[0], certificate=my_client_cert[2])
+
+        self.step(23)
         await cr1_cmd.send_remove_client_command(ccdid=response.ccdid)
```
[the remark this answers is not in the document]
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py]: Fixed

File: examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp (hunk 2 above)
Comment (gmarcosb) [reply examples/all-clusters-app/all-clusters-common/src/tls-certificate-management-instance.cpp:95]: I wasn't sure what to do with memory management (given how large the certs are - too big for stack) & wanted to leave memory allocation to the implementation

That's why there's the split between the server & `CertificateTableImpl`, with `TlsCertificateManagementCommandDelegate` basically taking care of the memory management before it then calls the underlying worker

File: src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp (hunk 3 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp:25]: Definitely, tracked @ https://github.com/project-chip/connectedhomeip/issues/40507 (which was going to be in this PR but got pushed out to followup#2)

File: src/python_testing/TC_TLSCERT.py (hunk 4 above)
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py:74]: Sorry that's a little confusing, `self.key` is in `class CertWithKey` which is populated with the value from this method

File: src/python_testing/TC_TLSCERT.py (hunk 8 above)
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py]: Thank you, I wasn't sure what I was supposed to stuff from the CSR & figured it probably didn't matter too much as the key is the important part 🤣 fixed

File: src/python_testing/TC_TLSCERT.py [hunk 14]
```diff
@@ -222,23 +237,23 @@ async def send_remove_root_command(
...
                                                          endpoint=self.endpoint, dev_ctrl=self.dev_ctrl, node_id=self.node_id, payloadCapability=ChipDeviceCtrl.TransportPayloadCapability.LARGE_PAYLOAD)
 
-                asserts.assert_true(type_matches(result, Clusters.TlsCertificateManagement.Commands.TLSClientCSRResponse),
-                                    "Unexpected return type for TLSClientCSR")
+                asserts.assert_true(type_matches(result, Clusters.TlsCertificateManagement.Commands.ClientCSRResponse),
+                                    "Unexpected return type for ClientCSR")
                 return result
             except InteractionModelError as e:
                 asserts.assert_equal(e.status, expected_status, "Unexpected error returned")
                 return e
```
[the remark this answers is not in the document]
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py]: Fixed

File: src/python_testing/TC_TLSCERT.py (hunk 7 above)
Comment (gmarcosb) [reply src/python_testing/TC_TLSCERT.py:558]: Added, hopefully got the right ones (thanks!), guessed at the `.S` prefix as it seemed to be what everything did

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter
Comment (gmarcosb) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:1]: Codegen re-run

File: src/python_testing/TC_TLSCERT.py [hunk 15]
```diff
@@ -275,295 +292,468 @@ async def read_tls_cert_attribute(self, attribute):
...
+        asserts.assert_equal(find_response.certificateDetails[0].caid, my_caid[1])
+
+        self.step(21)
+        await cr1_cmd.send_remove_root_command(caid=my_caid[0])
+
+        self.step(22)
+        await cr1_cmd.send_remove_fabric_command(cr2_cmd.fabric_index)
+
+    def desc_TC_TLSCERT_2_9(self) -> str:
         return "[TC-TLSCERT-3.1.8] ProvisionClientCertificate command verification"
```
Comment (robszewczyk) [inline src/python_testing/TC_TLSCERT.py]:
```suggestion
        return "[TC-TLSCERT-2.9] ProvisionClientCertificate command verification"
```

File: src/python_testing/TC_TLSCERT.py [hunk 16]
```diff
@@ -275,295 +292,468 @@ async def read_tls_cert_attribute(self, attribute):
...
+                     test_plan_support.verify_success()),
+            TestStep(24, test_plan_support.remove_fabric('CR2', 'CR1'), test_plan_support.verify_success()),
         ]
         return steps
 
     @run_if_endpoint_matches(has_cluster(Clusters.TlsCertificateManagement))
-    async def test_TC_TLSCERT_3_1_8(self):
-        endpoint = self.get_endpoint(default=1)
+    async def test_TC_TLSCERT_2_9(self):
         attributes = Clusters.TlsCertificateManagement.Attributes
```
Comment (robszewczyk) [inline src/python_testing/TC_TLSCERT.py]:
```suggestion
```

**Comment by gmarcosb** [comment]
Adding some context here from offline discussion with @andreilitvin @robszewczyk & @cecille:

The failing CI here seems to be due to some caching, where https://github.com/project-chip/connectedhomeip/pull/40817 is partially compiled but something is still using the old code prior to that PR

Tried using `no-ccache` both in description & in a commit, but it didn't help

My guess is that this output is where the stale cache is coming from:
```
Cache restored successfully
Cache restored from key: Linux-X64-linux-9e80867aec1a9cd9c7ba57fb7f16d4b95d3b8c24ef878bd41533ff3a6065b614
```

**Review by robszewczyk — APPROVED** [review]

File: src/python_testing/TC_PAVSTI_Utils.py [hunk 17]
```diff
@@ -150,11 +150,8 @@ def assert_valid_ccdid(self, ccdid: int) -> None:
         asserts.assert_greater_equal(ccdid, 0, "Invalid CCDID returned")
         asserts.assert_less_equal(ccdid, 65534, "Invalid CCDID returned")
 
-    def assert_valid_csr(
```
Comment (gmarcosb) [inline src/python_testing/TC_PAVSTI_Utils.py:153]: This should've been pulled out to a common utility instead of copy-pasted

Comment (gmarcosb) [reply src/python_testing/TC_PAVSTI_Utils.py:153]: Tracked for follow-up @ https://github.com/project-chip/connectedhomeip/issues/40931

File: src/python_testing/TC_TLSCERT.py [hunk 18]
```diff
@@ -275,295 +292,467 @@ async def read_tls_cert_attribute(self, attribute):
...
-            attribute_certs = await cr2_cmd.read_tls_cert_attribute(attribute=attributes.ProvisionedClientCertificates)
-            asserts.assert_greater_equal(len(attribute_certs), 1, "Expected at least 1 certificate")
-            found_certs = dict()
-            # Python controller is using TCP, which supports large payloads; need a way to test
-            # over MRP that certificate is not present
-            for cert in attribute_certs:
-                found_certs[cert.ccdid] = cert.clientCertificate
-            asserts.assert_in(response2.ccdid, found_certs, "ProvisionedClientCertificates should contain provisioned client cert")
-            asserts.assert_equal(found_certs[response2.ccdid], second_cert, "Expected matching certificate detail")
+        # TODO(gmarcosb): There be dragons with large attributes over large transport
```
Comment (gmarcosb) [inline src/python_testing/TC_TLSCERT.py:666]: Need to re-enable these test steps once large message transport for attributes is fixed

**Review by pidarped — APPROVED** [review]

File: src/app/BufferedReadCallback.cpp [hunk 19]
```diff
@@ -120,7 +120,12 @@ CHIP_ERROR BufferedReadCallback::BufferListItem(TLV::TLVReader & reader)
...
+    // Support chained buffer
```
Comment (bzbarsky-apple) [inline src/app/BufferedReadCallback.cpp:124]: The comments earlier in this method no longer quite make sense with this change.

But worse, when chained buffers end up being used, how does this work, exactly?  `GenerateListTLV` will try to use a `PacketBufferTLVReader` to read.  That will only read from the first chunk in the chain, no?  If the list item spanned multiple chunks, it's really not clear to me what will happen, exactly, in terms of whether you get errors or whether you get data corruption.

This core behavior change should have been in a separate PR, with unit tests covering it and review from people who actually understand this code.

That said, I assume the Python tests here are exercising this codepath... so how are they passing?  What am I missing?

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp [hunk 20]
```diff
@@ -327,11 +327,11 @@ void TlsCertificateManagementServer::HandleRemoveRootCertificate(HandlerContext
...
 }
 
-void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ctx, const TLSClientCSR::DecodableType & req)
+void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ctx, const ClientCSR::DecodableType & req)
 {
-    ChipLogDetail(Zcl, "TlsCertificateManagement: TLSClientCSR");
+    ChipLogDetail(Zcl, "TlsCertificateManagement: ClientCSR");
 
-    VerifyOrReturn(req.nonce.size() <= 128, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    VerifyOrReturn(req.nonce.size() <= kMaxNonceBytes, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
```
Comment (bzbarsky-apple) [inline src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:334]: This does not match the spec.  And presumably we should have cert tests covering this; if we don't or it's not part of the test plan, that should get fixed.

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp [hunk 21]
```diff
@@ -343,7 +343,8 @@ void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ct
     auto status = mDelegate.GenerateClientCsr(ctx.mRequestPath.mEndpointId, fabric, req, [&](auto & response) -> Status {
         VerifyOrDieWithMsg(response.ccdid <= kMaxClientCertId, NotSpecified, "Spec requires CCDID to be <= kMaxClientCertId");
         VerifyOrDieWithMsg(response.csr.size() <= 3000, NotSpecified, "Spec requires csr.size() to be <= 3000");
-        VerifyOrDieWithMsg(response.nonce.size() <= 128, NotSpecified, "Spec requires nonce.size() to be < kMaxClientCertId");
+        VerifyOrDieWithMsg(response.nonceSignature.size() <= kMaxSignatureBytes, NotSpecified,
```
Comment (bzbarsky-apple) [inline src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:346]: This does match the spec, but the spec looks fishy to me.  A ec-signature in our current crypto setup is always exactly 64 bytes, no?  So why is the constraint "max 128"?  Is that to allow for crypto agility, or something else?

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp [hunk 22]
```diff
@@ -360,21 +361,17 @@ void TlsCertificateManagementServer::HandleProvisionClientCertificate(HandlerCon
...
 
     VerifyOrReturn(req.ccdid <= kMaxClientCertId, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
-    const auto & detail = req.clientCertificateDetails;
-    VerifyOrReturn(!detail.clientCertificate.HasValue() || detail.clientCertificate.Value().size() <= kSpecMaxCertBytes,
+    VerifyOrReturn(req.clientCertificate.size() <= kSpecMaxCertBytes,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    size_t intermediateSize;
+    VerifyOrReturn(req.intermediateCertificates.ComputeSize(&intermediateSize) == CHIP_NO_ERROR,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidCommand));
+    VerifyOrReturn(intermediateSize <= kMaxIntermediateCertificates,
```
Comment (bzbarsky-apple) [inline src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:369]: I'm wondering why the spec says "0 to 10" instead of the more idiomatic "max 10", but ok.

File: src/app/BufferedReadCallback.cpp (hunk 19 above)
Comment (gmarcosb) [reply src/app/BufferedReadCallback.cpp:124]: Oh well that explains why even after I made this change this doesn't work 🤣

I was failing here in the TC test; then, when I fixed this, I was failing with "tlv buffer cut short", which pretty much sounds like what you're saying

I figured this change didn't make things worse but it also didn't fix things; this is tracked @ https://github.com/project-chip/connectedhomeip/issues/40938 (working on it now)

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 22 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:369]: That's my fault 🤣 it used to say "1 to 10", which was wrong, details @ https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/12201

I'll do follow-up to spec to just say "max 10" (but will have no effect on code)

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 20 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:334]: `kMaxNonceBytes=32`, per spec right?

(TC tests for edge cases, i.e. all failure testing tracked @ https://github.com/project-chip/connectedhomeip/issues/40907)

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 21 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:346]: No idea why, I just copied what it says 🤣 I thought maybe it was copied from `AttestationSignature`, but that's 64, so agreed 128 here makes no sense

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 20 above)
Comment (bzbarsky-apple) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:334]: > kMaxNonceBytes=32, per spec right?

Yes, but the spec constraint is "32", not "max 32".

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 21 above)
Comment (bzbarsky-apple) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:346]: Like I said: this matches the spec, but the spec looks weird.

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 20 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:334]: Ahh, woops! Will fix in follow-up

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 21 above)
Comment (bzbarsky-apple) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:346]: Aron confirmed this is a crypto agility affordance.

Comment (aronrosenberg) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:346]: This was done for future crypto agility reasons.

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 20 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:334]: @ https://github.com/project-chip/connectedhomeip/pull/40950

File: src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp (hunk 22 above)
Comment (gmarcosb) [reply src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp:369]: @ https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/12215

File: src/app/BufferedReadCallback.cpp (hunk 19 above)
Comment (pidarped) [reply src/app/BufferedReadCallback.cpp:124]: The issue is the allocation of a PacketBuffer of size `kMaxSecureSduLengthBytes`. If the underlying transport allows large payload, a buffer of capacity `kMaxLargeSecureSduLengthBytes` needs to be allocated, similar to what the CommandSender does. 
IIRC, there are checks in the code that  dis-allow chained buffers.

Comment (bzbarsky-apple) [reply src/app/BufferedReadCallback.cpp:124]: Yes, allocating a bigger buffer here would work better.
